### PR TITLE
KTOR-5567 Documentation for regex patterns in routing

### PR DESCRIPTION
**Pull request by marychatte**
[Design](https://youtrack.jetbrains.com/issue/KTOR-5279/Design-for-regex-patterns-support-in-routing)

### CODE REVIEW

File: topics/Routing_in_Ktor.md [hunk 1]
```diff
@@ -122,6 +124,54 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+import io.ktor.server.routing.*
+import io.ktor.server.response.*
+
+routing {
+    get(Regex("hello/[a-z]+")) {
+        call.respondText("Hello")
+    }
+}
+```
+Also, if the path matches by regex, it should start with a forward slash or nothing after the part consumed by regex. For example, path pattern `get(Regex("[a-z]+"))` will not match the path "hello1"
```
Comment (rsinukov) [inline topics/Routing_in_Ktor.md:170]: It was difficult for me to understand the meaning here. Maybe rephrase it, saying that the whole path segment should be consumed?

**Review by rsinukov — CHANGES_REQUESTED** [review]

File: topics/Routing_in_Ktor.md (hunk 1 above)
Comment (marychatte) [reply topics/Routing_in_Ktor.md:170]: Thanks, so I will change it for the next version:
```
Also, the whole path segment should be consumed by regex. For example, path pattern `get(Regex("[a-z]+"))` will not match the path `"hello1"` 
but will match the part `hello` of the path `hello/1` and leave `/1` for the next route.
```
Is it okay?

File: topics/Routing_in_Ktor.md [hunk 2]
```diff
@@ -122,6 +124,54 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+routing {
+    get(Regex(".+/hello")) {
+        call.respondText("Hello")
+    }
+}
+```
+
+#### Named groups
+
+To declare a named group in the regular expression, use `(?<name>...)` syntax.
```
Comment (e5l) [inline topics/Routing_in_Ktor.md:144]: please define what the named group is and why it is needed

File: topics/Routing_in_Ktor.md [hunk 3]
```diff
@@ -122,6 +124,54 @@ To access a parameter value inside the route handler, use the `call.parameters`
 A path parameter with a tailcard (`{param...}`) matches all the rest of the URL path and puts multiple values for each path segment into parameters using `param` as a key. For example, `/user/{param...}` matches `/user/john/settings`.
 To access path segments' values inside the route handler, use `call.parameters.getAll("param")`. For the example above, the `getAll` function will return an array containing the _john_ and _settings_ values.
 
+### Regular expression {id="regular_expression"}
+
+Regular expressions can be used with all defining route handlers functions: `route`, `get`, `post`, and so on. For example:
```
Comment (e5l) [inline topics/Routing_in_Ktor.md:129]: It would be great to add a link to the syntax of used regular expressions directly here.

File: topics/Routing_in_Ktor.md [hunk 4]
```diff
@@ -122,6 +124,54 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+
```
Comment (e5l) [inline topics/Routing_in_Ktor.md:130]: Please define a problem which you're going to solve with a sample: `Let's write a route matches blahblah`

File: topics/Routing_in_Ktor.md [hunk 5]
```diff
@@ -122,6 +124,54 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+import io.ktor.server.response.*
+
+routing {
+    get(Regex(".+/hello")) {
+        call.respondText("Hello")
+    }
+}
+```
+
+#### Named groups
```
Comment (e5l) [inline topics/Routing_in_Ktor.md:142]: I would name it using the purpose: `Accessing path parts in handler` and next explain how to use groups for that

**Review by e5l — CHANGES_REQUESTED** [review]

**Comment by marychatte** [comment]
@rsinukov and @e5l, thanks for your comments. I tried to fix all the problems. Could you please recheck it?

File: topics/Routing_in_Ktor.md [hunk 6]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+        call.respondText("Hello")
+    }
+}
+```
+With this route definition, any incoming request to a path ending with `/hello`, such as `/foo/hello`, `/bar/baz/hello`, 
+and so on, will be matched.
+
+#### Accessing path parts in handler
+
+In regular expressions, named groups are a way to capture a specific part of a string that matches a pattern and assign it a name.
```
Comment (e5l) [inline topics/Routing_in_Ktor.md:150]: is `,` needed?

**Review by e5l — APPROVED** [review]
Looks better, lgtm

File: topics/Routing_in_Ktor.md (hunk 6 above)
Comment (marychatte) [reply topics/Routing_in_Ktor.md:150]: It is what `Grammarly` recommends to do :)
As I understand, you can add a comma or not for short introductory phrases

Comment (e5l) [reply topics/Routing_in_Ktor.md:150]: got it, thanks

File: topics/Routing_in_Ktor.md [hunk 7]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
 A path parameter with a tailcard (`{param...}`) matches all the rest of the URL path and puts multiple values for each path segment into parameters using `param` as a key. For example, `/user/{param...}` matches `/user/john/settings`.
 To access path segments' values inside the route handler, use `call.parameters.getAll("param")`. For the example above, the `getAll` function will return an array containing the _john_ and _settings_ values.
 
+### Regular expression {id="regular_expression"}
+
+Regular expressions can be used with all defining route handlers functions: `route`, `get`, `post`, and so on. 
+
+> To know more about regular expressions, see [Kotlin documentation](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/-regex/).
+
+Let's write a route matches any path that ends with `/hello`.
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:133]: **that** matches

File: topics/Routing_in_Ktor.md [hunk 8]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+
+```kotlin
+import io.ktor.server.routing.*
+import io.ktor.server.response.*
+
+routing {
+    get(Regex(".+/hello")) {
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:140]: It would be helpful to mention that a regex of a route is implicitly surrounded by `^` and `$`.

File: topics/Routing_in_Ktor.md [hunk 9]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+}
+```
+With this route definition, any incoming request to a path ending with `/hello`, such as `/foo/hello`, `/bar/baz/hello`, 
+and so on, will be matched.
+
+#### Accessing path parts in handler
+
+In regular expressions, named groups are a way to capture a specific part of a string that matches a pattern and assign it a name.
+Named groups are defined using the following syntax `(?<name>pattern)`, where `name` is the name of the group and
+`pattern` is the regular expression pattern that matches the group.
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:152]: * using the `(?<name>pattern)` syntax
* is the name of **a** group
* `pattern` is **a** regular expression pattern that matches the group

File: topics/Routing_in_Ktor.md [hunk 10]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+
+By defining a named group in the route function, you can capture a part of the path, and then in the handler function,
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:154]: in *a* route function

File: topics/Routing_in_Ktor.md [hunk 11]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+you can access the captured parameter using the `call.parameters` property.
+
+For example, you can define a route that matches requests to a path that includes an integer identifier followed by `/hello`.
+
+```kotlin
+import io.ktor.server.routing.*
+import io.ktor.server.response.*
+
+routing {
+    get(Regex("(?<id>\\d+)/hello")) {
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:164]: Maybe we should use triple quotes in the examples to avoid double backslashes.

File: topics/Routing_in_Ktor.md [hunk 12]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+        val id = call.parameters["id"]!!
+        call.respondText(id)
+    }
+}
+```
+In the code below, the `(?<id>\\d+)` named group is used to capture the integer identifier `id` from the path,
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:170]: from a requested path

File: topics/Routing_in_Ktor.md [hunk 13]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+and so on, will be matched.
+
+#### Accessing path parts in handler
+
+In regular expressions, named groups are a way to capture a specific part of a string that matches a pattern and assign it a name.
+Named groups are defined using the following syntax `(?<name>pattern)`, where `name` is the name of the group and
+`pattern` is the regular expression pattern that matches the group.
+
+By defining a named group in the route function, you can capture a part of the path, and then in the handler function,
+you can access the captured parameter using the `call.parameters` property.
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:155]: It's better to use the word `object` instead of `property` because it doesn't matter what the relationship is between `ApplicationCall` and `parameters`.

File: topics/Routing_in_Ktor.md [hunk 14]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+
+routing {
+    get(Regex("(?<id>\\d+)/hello")) {
+        val id = call.parameters["id"]!!
+        call.respondText(id)
+    }
+}
+```
+In the code below, the `(?<id>\\d+)` named group is used to capture the integer identifier `id` from the path,
+and the `call.parameters` property is used to access the captured id parameter in the handler function.
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:171]: id in backticks

File: topics/Routing_in_Ktor.md [hunk 15]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+
+Unnamed groups can't be accessed inside the route handler, but you can use them to match the path. For example, path `hello/world` will be matched while `hello/World` not:
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:173]: Unnamed groups can't be accessed inside **a regex route handler**
There are no unnamed groups in the following example.

File: topics/Routing_in_Ktor.md [hunk 16]
```diff
@@ -122,6 +124,67 @@ To access a parameter value inside the route handler, use the `call.parameters`
...
+import io.ktor.server.routing.*
+import io.ktor.server.response.*
+
+routing {
+    get(Regex("hello/[a-z]+")) {
+        call.respondText("Hello")
+    }
+}
+```
+Also, the whole path segment should be consumed by regex. For example, path pattern `get(Regex("[a-z]+"))` will not match the path `"hello1"`
```
Comment (Stexxe) [inline topics/Routing_in_Ktor.md:185]: What do you mean here by **should**? Is it consumed or not?

File: topics/Routing_in_Ktor.md (hunk 8 above)
Comment (marychatte) [reply topics/Routing_in_Ktor.md:140]: It doesn't work for the route function. In next example `Regex(".+/hello")` wouldn't end with `$`:
```
route(Regex(".+/hello")) {
    get(Regex("/world") {
    ...
    }
}
```
